[Cocoa] Refactor: remote MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification()

jernoble opened the pull request:
#### 58e56ea67c7b543bbec9cf17ffcabab78b7d6c2e
<pre>
[Cocoa] Refactor: remote MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification()
<a href="https://bugs.webkit.org/show_bug.cgi?id=264746">https://bugs.webkit.org/show_bug.cgi?id=264746</a>
<a href="https://rdar.apple.com/118335975">rdar://118335975</a>

Reviewed by Andy Estes.

MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification() supported an elaborate list of
notification types, only two of which were used after the removal of MediaPlayerPrivateAVFoundationCF.
Replace these two notifications with standard callbacks.

* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::MediaPlayerPrivateAVFoundation):
(WebCore::MediaPlayerPrivateAVFoundation::setPreload):
(WebCore::MediaPlayerPrivateAVFoundation::trackModeChanged):
(WebCore::MediaPlayerPrivateAVFoundation::setDelayCallbacks const): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::mainThreadCallback): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::clearMainThreadPendingFlag): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::dispatchNotification): Deleted.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVAssetForURL):
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayer):
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayerItem):
(WebCore::MediaPlayerPrivateAVFoundationObjC::checkPlayability):
(WebCore::MediaPlayerPrivateAVFoundationObjC::seekToTargetInternal):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setPlayerRate):
(WebCore::MediaPlayerPrivateAVFoundationObjC::paintCurrentFrameInContext):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setWirelessVideoPlaybackDisabled):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setShouldPlayToPlaybackTarget):

Canonical link: <a href="https://commits.webkit.org/270659@main">https://commits.webkit.org/270659@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b0e2f5c2907b7294e4471a23f5405d45b3d46f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26012 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4624 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27295 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28112 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23826 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26330 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6387 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2053 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23895 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22413 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28692 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23365 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29426 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23742 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27304 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3155 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1365 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4546 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6266 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3614 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3475 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/219af97bcc738a1e0f92e225b0cf973a3bcb81f9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/20419

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2b0e2f5c2907b7294e4471a23f5405d45b3d46f6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/20419

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -1255,12 +1244,14 @@ static URL conformFragmentIdentifierForURL(const URL& url)
 
     INFO_LOG(LOGIDENTIFIER);
-    WeakPtr weakThis { *this };
+    __block WeakPtr weakThis { *this };
```
Nit: should be able to just create the WeakPtr in the capture list (i.e. `weakThis = WeakPtr { *this }`)

jernoble replied:
I don't think you can do that; the outer lambda is an Objective-c block, not a C++ lambda.

aestes replied:
Oops, I overlooked the outer block.

aestes commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -1255,12 +1244,14 @@ static URL conformFragmentIdentifierForURL(const URL& url)
     [m_avAsset loadValuesAsynchronouslyForKeys:@[@"playable", @"tracks"] completionHandler:^{
-        callOnMainThread([weakThis] {
+        ensureOnMainThread([weakThis = WTFMove(weakThis)] {
+            if (weakThis)
```
Nit: Simpler to early return?
```
if (!weakThis)
    return;
```

jernoble committed:
[Cocoa] Refactor: remote MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification()
https://bugs.webkit.org/show_bug.cgi?id=264746
rdar://118335975

Reviewed by Andy Estes.

MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification() supported an elaborate list of
notification types, only two of which were used after the removal of MediaPlayerPrivateAVFoundationCF.
Replace these two notifications with standard callbacks.

* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::MediaPlayerPrivateAVFoundation):
(WebCore::MediaPlayerPrivateAVFoundation::setPreload):
(WebCore::MediaPlayerPrivateAVFoundation::trackModeChanged):
(WebCore::MediaPlayerPrivateAVFoundation::setDelayCallbacks const): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::mainThreadCallback): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::clearMainThreadPendingFlag): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::scheduleMainThreadNotification): Deleted.
(WebCore::MediaPlayerPrivateAVFoundation::dispatchNotification): Deleted.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVAssetForURL):
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayer):
(WebCore::MediaPlayerPrivateAVFoundationObjC::createAVPlayerItem):
(WebCore::MediaPlayerPrivateAVFoundationObjC::checkPlayability):
(WebCore::MediaPlayerPrivateAVFoundationObjC::seekToTargetInternal):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setPlayerRate):
(WebCore::MediaPlayerPrivateAVFoundationObjC::paintCurrentFrameInContext):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setWirelessVideoPlaybackDisabled):
(WebCore::MediaPlayerPrivateAVFoundationObjC::setShouldPlayToPlaybackTarget):

Canonical link: https://commits.webkit.org/270659@main

webkit-commit-queue commented:
Committed 270659@main (58e56ea67c7b): <https://commits.webkit.org/270659@main>

Reviewed commits have been landed. Closing PR #20419 and removing active labels.
